src/backend/core/resource_server/auth.py

```diff
@@ -0,0 +1,118 @@
+"""Resource Server authentication class"""
+
+from base64 import b64decode
+
+from requests.exceptions import HTTPError
+from rest_framework.authentication import BaseAuthentication, get_authorization_header
+from rest_framework.exceptions import AuthenticationFailed
+
+from ..models import User
+from .clients import AuthorizationServerClient, ResourceServerClient
+from .utils import get_settings
+
+
+class ResourceServerAuthentication(BaseAuthentication):
+    """Token-based authentication for Resource Server (RS).
+
+    Authenticate by passing the token received from the OIDC Provider (OP).
+    The Resource Server will introspect the token, while the OIDC Provider validates
+    its integrity and permissions.
+    """
+
+    @staticmethod
+    def _decode_authorization_header(request):
+        """Get token and received_secret passed by the Service Provider (SP)."""
+
+        authorization_header = get_authorization_header(request).split()
+
+        if (
+            not authorization_header
+            or authorization_header[0].lower() != "Bearer".lower().encode()
+        ):
+            msg = "Invalid token header. No credentials provided."
+            raise AuthenticationFailed(msg)
+
+        if len(authorization_header) == 1:
+            msg = "Invalid token header. No credentials provided."
+            raise AuthenticationFailed(msg)
+
+        if len(authorization_header) > 2:
+            msg = "Invalid token header. Token string should not contain spaces."
+            raise AuthenticationFailed(msg)
+
+        decoded_bearer = b64decode(authorization_header[1]).decode("utf-8")
+
+        # FIXME: inherited from France Connect mocked RS, bad practice, a bearer token should respect format specified in the RFC
+        authorization_data = decoded_bearer.split(":")
+        if len(authorization_data) != 2:
+            msg = "Token should contain a token and a received_secret."
+            raise AuthenticationFailed(msg)
+
+        token, received_secret = authorization_data
+
+        return (token, received_secret)
+
+    @staticmethod
+    def authenticate_service_provider(received_secret):
+        """Authenticate the Service Provider (SP) with a shared secret.
+
+        This method is temporary, and inspired by Agent Connect mocked resource servers.
+        """
+        if received_secret != get_settings("OIDC_RS_AUTH_SECRET"):
```

Comment:
does it mean that every service provider that will want to get data from our resource server will need to be registered in this setting?

Reply:
Hmm, not exactly.

This part of the design is flawed and originates from the Agent Connect development stack. It needs improvements.

Currently, the service provider and the resource server share a secret. This secret is an environment setting in Agent Connect dev stack, and shared by several service providers for the sake of simplicity. 

In the current implementation, the shared secret is compared with a constant setting. While this adds an extra layer of security, it can be easily bypassed. This check was copied from the Agent Connect data provider mocks.

I suggest releasing an initial version of the code without this check, and receiving an RFC-compliant bearer token. We can then iterate to add a proper database model that stores client_id and client_secret pairs for the service providers interacting with Desk, if needed.

Reply:
Note: The service provider passes an authorization header to the resource server mocks, which contains a bearer token.

However, in the current implementation on the Agent Connect side, the bearer token is non-compliant with RFC standards. They combine the access_token value and the shared secret between the service provider and the resource server into a single bearer token. This results in a token formatted as `Bearer actualBearerTokenValue:sharedSecret`. On the resource server side, we then need to split this bearer token into two parts, which is also non-compliant with RFC standards. This part needs to be refactored!